File: KeplerSynapseNet/crush-main/internal/patch/patch.go
package patch

import (
	"errors"
	"fmt"
	"strconv"
	"strings"
)

type LineKind int

const (
	LineContext LineKind = iota
	LineAdd
	LineDelete
)

type HunkLine struct {
	Kind LineKind
	Text string
}

type Hunk struct {
	OldStart int
	OldCount int
	NewStart int
	NewCount int
	Lines    []HunkLine
}

type FilePatch struct {
	OldPath string
	NewPath string
	Hunks   []Hunk
}

func ParseUnifiedDiff(input string) ([]FilePatch, error) {
	diffText := extractDiffBlock(input)
	lines := splitLines(diffText)

	var patches []FilePatch
	var cur *FilePatch
	var curHunk *Hunk

	flushHunk := func() {
		if cur == nil || curHunk == nil {
			return
		}
		cur.Hunks = append(cur.Hunks, *curHunk)
		curHunk = nil
	}
	flushFile := func() {
		if cur == nil {
			return
		}
		flushHunk()
		if cur.OldPath != "" || cur.NewPath != "" {
			patches = append(patches, *cur)
		}
		cur = nil
	}

	for _, raw := range lines {
		line := strings.TrimRight(raw, "\r")
		if strings.HasPrefix(line, "diff --git ") {
			flushFile()
			oldPath, newPath := parseDiffGitLine(line)
			cur = &FilePatch{
				OldPath: oldPath,
				NewPath: newPath,
			}
			continue
		}

		if strings.HasPrefix(line, "--- ") {
			if cur == nil {
				cur = &FilePatch{}
			}
			cur.OldPath = normalizePath(strings.TrimSpace(strings.TrimPrefix(line, "--- ")))
			continue
		}
		if strings.HasPrefix(line, "+++ ") {
			if cur == nil {
				cur = &FilePatch{}
			}
			cur.NewPath = normalizePath(strings.TrimSpace(strings.TrimPrefix(line, "+++ ")))
			continue
		}

		if strings.HasPrefix(line, "@@ ") {
			if cur == nil {
				cur = &FilePatch{}
			}
			flushHunk()
			h, err := parseHunkHeader(line)
			if err != nil {
				return nil, err
			}
			curHunk = &h
			continue
		}

		if curHunk != nil {
			if line == "\\ No newline at end of file" {
				continue
			}
			if line == "" {
				curHunk.Lines = append(curHunk.Lines, HunkLine{Kind: LineContext, Text: ""})
				continue
			}
			switch line[0] {
			case ' ':
				curHunk.Lines = append(curHunk.Lines, HunkLine{Kind: LineContext, Text: line[1:]})
			case '+':
				if strings.HasPrefix(line, "+++") {
					continue
				}
				curHunk.Lines = append(curHunk.Lines, HunkLine{Kind: LineAdd, Text: line[1:]})
			case '-':
				if strings.HasPrefix(line, "---") {
					continue
				}
				curHunk.Lines = append(curHunk.Lines, HunkLine{Kind: LineDelete, Text: line[1:]})
			default:
				return nil, fmt.Errorf("invalid hunk line: %q", line)
			}
			continue
		}
	}

	flushFile()

	if len(patches) == 0 {
		return nil, errors.New("no file diffs found")
	}
	for _, p := range patches {
		if len(p.Hunks) == 0 && p.OldPath != "/dev/null" && p.NewPath != "/dev/null" {
			return nil, errors.New("no hunks found")
		}
	}

	return patches, nil
}

func Apply(oldContent string, p FilePatch) (string, error) {
	hasFinalNewline := strings.HasSuffix(oldContent, "\n")
	oldLines := splitContentLines(oldContent)

	out := make([]string, 0, len(oldLines))
	oldIdx := 0

	for _, h := range p.Hunks {
		target := h.OldStart - 1
		if target < 0 {
			target = 0
		}
		if target > len(oldLines) {
			return "", fmt.Errorf("hunk starts beyond end of file: %d", h.OldStart)
		}
		out = append(out, oldLines[oldIdx:target]...)
		oldIdx = target

		for _, hl := range h.Lines {
			switch hl.Kind {
			case LineContext:
				if oldIdx >= len(oldLines) {
					return "", errors.New("context exceeds file length")
				}
				if oldLines[oldIdx] != hl.Text {
					return "", fmt.Errorf("context mismatch at line %d", oldIdx+1)
				}
				out = append(out, hl.Text)
				oldIdx++
			case LineDelete:
				if oldIdx >= len(oldLines) {
					return "", errors.New("delete exceeds file length")
				}
				if oldLines[oldIdx] != hl.Text {
					return "", fmt.Errorf("delete mismatch at line %d", oldIdx+1)
				}
				oldIdx++
			case LineAdd:
				out = append(out, hl.Text)
			default:
				return "", errors.New("unknown line kind")
			}
		}
	}

	out = append(out, oldLines[oldIdx:]...)
	newContent := strings.Join(out, "\n")
	if hasFinalNewline || oldContent == "" {
		newContent += "\n"
	}
	return newContent, nil
}

func extractDiffBlock(input string) string {
	trimmed := strings.TrimSpace(input)
	if trimmed == "" {
		return ""
	}

	lines := splitLines(trimmed)
	best := ""

	inFence := false
	fenceStart := -1
	for i, line := range lines {
		l := strings.TrimSpace(line)
		if strings.HasPrefix(l, "```") {
			if !inFence {
				inFence = true
				fenceStart = i + 1
				continue
			}
			inFence = false
			if fenceStart >= 0 && fenceStart < i {
				block := strings.Join(lines[fenceStart:i], "\n")
				if strings.Contains(block, "diff --git ") || (strings.Contains(block, "--- ") && strings.Contains(block, "+++ ")) {
					return block
				}
				if best == "" {
					best = block
				}
			}
			fenceStart = -1
		}
	}

	if best != "" {
		return best
	}
	return trimmed
}

func splitLines(s string) []string {
	if s == "" {
		return nil
	}
	return strings.Split(s, "\n")
}

func splitContentLines(content string) []string {
	if content == "" {
		return []string{}
	}
	c := strings.TrimSuffix(content, "\n")
	if c == "" {
		return []string{}
	}
	return strings.Split(c, "\n")
}

func parseDiffGitLine(line string) (string, string) {
	rest := strings.TrimSpace(strings.TrimPrefix(line, "diff --git "))
	parts := strings.Fields(rest)
	if len(parts) >= 2 {
		return normalizePath(parts[0]), normalizePath(parts[1])
	}
	return "", ""
}

func normalizePath(p string) string {
	p = strings.TrimSpace(p)
	if p == "" {
		return p
	}
	if p == "/dev/null" {
		return p
	}
	p = strings.TrimPrefix(p, "a/")
	p = strings.TrimPrefix(p, "b/")
	return p
}

func parseHunkHeader(line string) (Hunk, error) {
	if !strings.HasPrefix(line, "@@") {
		return Hunk{}, fmt.Errorf("invalid hunk header: %q", line)
	}
	end := strings.Index(line[2:], "@@")
	if end == -1 {
		return Hunk{}, fmt.Errorf("invalid hunk header: %q", line)
	}
	body := strings.TrimSpace(line[2 : 2+end])
	fields := strings.Fields(body)
	if len(fields) < 2 {
		return Hunk{}, fmt.Errorf("invalid hunk header: %q", line)
	}
	oldStart, oldCount, err := parseRange(fields[0])
	if err != nil {
		return Hunk{}, err
	}
	newStart, newCount, err := parseRange(fields[1])
	if err != nil {
		return Hunk{}, err
	}
	return Hunk{
		OldStart: oldStart,
		OldCount: oldCount,
		NewStart: newStart,
		NewCount: newCount,
		Lines:    nil,
	}, nil
}

func parseRange(tok string) (int, int, error) {
	if tok == "" {
		return 0, 0, errors.New("empty range")
	}
	sign := tok[0]
	if sign != '-' && sign != '+' {
		return 0, 0, fmt.Errorf("invalid range token: %q", tok)
	}
	tok = tok[1:]
	if tok == "" {
		return 0, 0, fmt.Errorf("invalid range token: %q", tok)
	}
	startStr, countStr, hasCount := strings.Cut(tok, ",")
	start, err := strconv.Atoi(startStr)
	if err != nil {
		return 0, 0, fmt.Errorf("invalid range start: %q", tok)
	}
	count := 1
	if hasCount {
		if countStr == "" {
			count = 0
		} else {
			count, err = strconv.Atoi(countStr)
			if err != nil {
				return 0, 0, fmt.Errorf("invalid range count: %q", tok)
			}
		}
	}
	return start, count, nil
}

File: KeplerSynapseNet/src/ide/patch.cpp
#include "ide/patch.h"

#include <sstream>

namespace synapse {
namespace ide {

namespace {

std::vector<std::string> splitLines(const std::string& s) {
    if (s.empty()) return {};
    std::vector<std::string> result;
    std::istringstream iss(s);
    std::string line;
    while (std::getline(iss, line)) {
        result.push_back(line);
    }
    return result;
}

std::vector<std::string> splitContentLines(const std::string& content) {
    if (content.empty()) return {};
    std::string c = content;
    if (!c.empty() && c.back() == '\n') c.pop_back();
    if (c.empty()) return {};
    return splitLines(c);
}

std::string extractDiffBlock(const std::string& input) {
    std::string trimmed = input;
    while (!trimmed.empty() && (trimmed.front() == ' ' || trimmed.front() == '\n' ||
                                 trimmed.front() == '\r' || trimmed.front() == '\t')) {
        trimmed.erase(trimmed.begin());
    }
    while (!trimmed.empty() && (trimmed.back() == ' ' || trimmed.back() == '\n' ||
                                 trimmed.back() == '\r' || trimmed.back() == '\t')) {
        trimmed.pop_back();
    }
    if (trimmed.empty()) return "";

    auto lines = splitLines(trimmed);
    std::string best;

    bool inFence = false;
    int fenceStart = -1;
    for (size_t i = 0; i < lines.size(); ++i) {
        std::string l = lines[i];
        auto pos = l.find_first_not_of(" \t");
        if (pos != std::string::npos) l = l.substr(pos);
        if (l.substr(0, 3) == "```") {
            if (!inFence) {
                inFence = true;
                fenceStart = static_cast<int>(i) + 1;
                continue;
            }
            inFence = false;
            if (fenceStart >= 0 && fenceStart < static_cast<int>(i)) {
                std::string block;
                for (int j = fenceStart; j < static_cast<int>(i); ++j) {
                    if (!block.empty()) block += "\n";
                    block += lines[j];
                }
                if (block.find("diff --git ") != std::string::npos ||
                    (block.find("--- ") != std::string::npos &&
                     block.find("+++ ") != std::string::npos)) {
                    return block;
                }
                if (best.empty()) best = block;
            }
            fenceStart = -1;
        }
    }

    if (!best.empty()) return best;
    return trimmed;
}

std::string normalizePath(const std::string& p) {
    std::string result = p;
    while (!result.empty() && (result.front() == ' ' || result.front() == '\t')) {
        result.erase(result.begin());
    }
    while (!result.empty() && (result.back() == ' ' || result.back() == '\t')) {
        result.pop_back();
    }
    if (result.empty()) return result;
    if (result == "/dev/null") return result;
    if (result.size() > 2 && result[0] == 'a' && result[1] == '/') result = result.substr(2);
    else if (result.size() > 2 && result[0] == 'b' && result[1] == '/') result = result.substr(2);
    return result;
}

void parseDiffGitLine(const std::string& line, std::string& oldPath, std::string& newPath) {
    std::string rest = line.substr(11);
    while (!rest.empty() && (rest.front() == ' ' || rest.front() == '\t')) {
        rest.erase(rest.begin());
    }
    auto spacePos = rest.find(' ');
    if (spacePos != std::string::npos) {
        oldPath = normalizePath(rest.substr(0, spacePos));
        newPath = normalizePath(rest.substr(spacePos + 1));
    }
}

bool parseRange(const std::string& tok, int& start, int& count) {
    if (tok.empty()) return false;
    char sign = tok[0];
    if (sign != '-' && sign != '+') return false;
    std::string rest = tok.substr(1);
    if (rest.empty()) return false;
    auto commaPos = rest.find(',');
    if (commaPos == std::string::npos) {
        try {
            start = std::stoi(rest);
            count = 1;
            return true;
        } catch (...) { return false; }
    }
    try {
        start = std::stoi(rest.substr(0, commaPos));
        std::string countStr = rest.substr(commaPos + 1);
        count = countStr.empty() ? 0 : std::stoi(countStr);
        return true;
    } catch (...) { return false; }
}

bool parseHunkHeader(const std::string& line, Hunk& h) {
    if (line.substr(0, 2) != "@@") return false;
    auto end = line.find("@@", 2);
    if (end == std::string::npos) return false;
    std::string body = line.substr(2, end - 2);
    while (!body.empty() && body.front() == ' ') body.erase(body.begin());
    while (!body.empty() && body.back() == ' ') body.pop_back();

    auto spacePos = body.find(' ');
    if (spacePos == std::string::npos) return false;
    std::string oldTok = body.substr(0, spacePos);
    std::string newTok = body.substr(spacePos + 1);
    while (!newTok.empty() && newTok.front() == ' ') newTok.erase(newTok.begin());
    auto spacePos2 = newTok.find(' ');
    if (spacePos2 != std::string::npos) newTok = newTok.substr(0, spacePos2);

    if (!parseRange(oldTok, h.oldStart, h.oldCount)) return false;
    if (!parseRange(newTok, h.newStart, h.newCount)) return false;
    return true;
}

}

std::vector<FilePatch> parseUnifiedDiff(const std::string& input, std::string& error) {
    std::string diffText = extractDiffBlock(input);
    auto lines = splitLines(diffText);

    std::vector<FilePatch> patches;
    FilePatch* cur = nullptr;
    Hunk* curHunk = nullptr;

    std::vector<FilePatch> storage;
    storage.reserve(16);

    auto flushHunk = [&]() {
        if (!cur || !curHunk) return;
        cur->hunks.push_back(*curHunk);
        curHunk = nullptr;
    };

    auto flushFile = [&]() {
        if (!cur) return;
        flushHunk();
        if (!cur->oldPath.empty() || !cur->newPath.empty()) {
            patches.push_back(*cur);
        }
        cur = nullptr;
    };

    Hunk tmpHunk;
    FilePatch tmpPatch;

    for (const auto& raw : lines) {
        std::string line = raw;
        if (!line.empty() && line.back() == '\r') line.pop_back();

        if (line.size() >= 11 && line.substr(0, 11) == "diff --git ") {
            flushFile();
            tmpPatch = FilePatch{};
            parseDiffGitLine(line, tmpPatch.oldPath, tmpPatch.newPath);
            storage.push_back(tmpPatch);
            cur = &storage.back();
            continue;
        }

        if (line.size() >= 4 && line.substr(0, 4) == "--- ") {
            if (!cur) {
                tmpPatch = FilePatch{};
                storage.push_back(tmpPatch);
                cur = &storage.back();
            }
            cur->oldPath = normalizePath(line.substr(4));
            continue;
        }
        if (line.size() >= 4 && line.substr(0, 4) == "+++ ") {
            if (!cur) {
                tmpPatch = FilePatch{};
                storage.push_back(tmpPatch);
                cur = &storage.back();
            }
            cur->newPath = normalizePath(line.substr(4));
            continue;
        }

        if (line.size() >= 3 && line.substr(0, 3) == "@@ ") {
            if (!cur) {
                tmpPatch = FilePatch{};
                storage.push_back(tmpPatch);
                cur = &storage.back();
            }
            flushHunk();
            tmpHunk = Hunk{};
            if (!parseHunkHeader(line, tmpHunk)) {
                error = "invalid hunk header: " + line;
                return {};
            }
            curHunk = &tmpHunk;
            continue;
        }

        if (curHunk) {
            if (line == "\\ No newline at end of file") continue;
            if (line.empty()) {
                curHunk->lines.push_back(HunkLine{LineKind::Context, ""});
                continue;
            }
            switch (line[0]) {
            case ' ':
                curHunk->lines.push_back(HunkLine{LineKind::Context, line.substr(1)});
                break;
            case '+':
                if (line.substr(0, 3) == "+++") continue;
                curHunk->lines.push_back(HunkLine{LineKind::Add, line.substr(1)});
                break;
            case '-':
                if (line.substr(0, 3) == "---") continue;
                curHunk->lines.push_back(HunkLine{LineKind::Delete, line.substr(1)});
                break;
            default:
                error = "invalid hunk line: " + line;
                return {};
            }
            continue;
        }
    }

    flushFile();

    if (patches.empty()) {
        error = "no file diffs found";
        return {};
    }
    for (const auto& p : patches) {
        if (p.hunks.empty() && p.oldPath != "/dev/null" && p.newPath != "/dev/null") {
            error = "no hunks found";
            return {};
        }
    }

    return patches;
}

std::string applyPatch(const std::string& oldContent, const FilePatch& patch,
                       std::string& error) {
    bool hasFinalNewline = !oldContent.empty() && oldContent.back() == '\n';
    auto oldLines = splitContentLines(oldContent);

    std::vector<std::string> out;
    out.reserve(oldLines.size());
    int oldIdx = 0;

    for (const auto& h : patch.hunks) {
        int target = h.oldStart - 1;
        if (target < 0) target = 0;
        if (target > static_cast<int>(oldLines.size())) {
            error = "hunk starts beyond end of file: " + std::to_string(h.oldStart);
            return "";
        }
        for (int i = oldIdx; i < target; ++i) {
            out.push_back(oldLines[i]);
        }
        oldIdx = target;

        for (const auto& hl : h.lines) {
            switch (hl.kind) {
            case LineKind::Context:
                if (oldIdx >= static_cast<int>(oldLines.size())) {
                    error = "context exceeds file length";
                    return "";
                }
                if (oldLines[oldIdx] != hl.text) {
                    error = "context mismatch at line " + std::to_string(oldIdx + 1);
                    return "";
                }
                out.push_back(hl.text);
                ++oldIdx;
                break;
            case LineKind::Delete:
                if (oldIdx >= static_cast<int>(oldLines.size())) {
                    error = "delete exceeds file length";
                    return "";
                }
                if (oldLines[oldIdx] != hl.text) {
                    error = "delete mismatch at line " + std::to_string(oldIdx + 1);
                    return "";
                }
                ++oldIdx;
                break;
            case LineKind::Add:
                out.push_back(hl.text);
                break;
            }
        }
    }

    for (int i = oldIdx; i < static_cast<int>(oldLines.size()); ++i) {
        out.push_back(oldLines[i]);
    }

    std::string newContent;
    for (size_t i = 0; i < out.size(); ++i) {
        if (i > 0) newContent += "\n";
        newContent += out[i];
    }
    if (hasFinalNewline || oldContent.empty()) {
        newContent += "\n";
    }

    return newContent;
}

}
}
